### ISSUE TITLE
Stuck on "Compiling to native code"

### ISSUE THREAD

**Issue by svnpenn** [red/red#3412]
Running either one of these hangs on the final step, while peaking the CPU core
in use. Must use Ctrl + C to stop:

~~~red
$ cat hello.red
Red [
   Title: "Simple hello world script"
]
print "Hello World!"
~~~

~~~
$ red-063 -r hello.red

-=== Red Compiler 0.6.3 ===-

Compiling C:\hello.red ...
...compilation time : 1534 ms

Target: MSDOS

Compiling to native code...
~~~

~~~
$ red-063 -c hello.red

-=== Red Compiler 0.6.3 ===-

Compiling C:\hello.red ...
Compiling libRedRT...
...compilation time : 2543 ms

Compiling to native code...
~~~

**Comment by svnpenn** [comment]
Ok so the compiler is working, its just incredibly slow:

<pre>
$ time red-063 -r hello.red
-=== Red Compiler 0.6.3 ===-

Compiling C:\hello.red ...
...compilation time : 1373 ms

Target: MSDOS

Compiling to native code...
...compilation time : <b>38735 ms</b>
...linking time     : 250 ms
...output file size : 595968 bytes
...output file      : C:\hello.exe
</pre>

Compare with GCC:

<pre>
$ cat hello.c
#include <stdio.h>
int main() {
  puts("hello world");
}

$ time x86_64-w64-mingw32-gcc hello.c

real    <b>0m0.234s</b>
user    0m0.076s
sys     0m0.106s
</pre>

**Comment by svnpenn** [comment]
@PeterWAWood are you certain of this - I am not seeing any "Red runtime" being compiled or saved, unless it is hidden or in another folder - do I need to be using some other option so that the Red Runtime is not compiled every time?

**Comment by PeterWAWood** [comment]
For completeness, these are the timings I got on a second compile of your program:
```text
time rebol -qws ../../Red/red/red.r hello.red
...
...using libRedRT built on 9-Jun-2018/11:29:29+8:00
...compilation time : 25 ms

Compiling to native code...
...compilation time : 1213 ms
...linking time     : 52 ms
...output file size : 77824 bytes

real	0m1.384s
user	0m1.319s
sys	0m0.022s
``` 

The current compiler is a throw-away bootstrap compiler at the Alpha stage. It shouldn't be expected that it will be as fast as GCC.

**Comment by PeterWAWood** [comment]
The `-r` option compiles the runtime and includes it in the executable. Use the `-c` option to use the libRedRT runtime. (If the compiler doesn't find a libRedRT in the source directory, it will compile that first before compiling the program).

**Comment by PeterWAWood** [comment]
Please could you ask such questions in the [Red Help Chat Room](https://gitter.im/red/help) or on the [Red Mailing List](https://groups.google.com/forum/#!forum/red-lang) next time you have questions about Red.

**Comment by svnpenn** [comment]
@PeterWAWood ok it is working for me now - however some problems:

if i use `-r` it makes a static build (like i want), but doesnt leave the static
library behind - so it has to be compiled each time

if i use `-c` it makes a shared build (like i dont want), but it leaves the
shared library behind - so it does not need to be compiled each time

can i use some option to resolve this

**Comment by PeterWAWood** [comment]
Please ask this question in either the Red Help Chat Room or on the Red Mailing List.

**Comment by dockimbel** [comment]
@svnpenn 
> Compare with GCC:

You are comparing apples with oranges, Red is a very high-level language, while C is a very low-level language. The work done by the compiler is very different in both cases, not even mentioning that the whole Red standard library is recompiled each time if you insist on using the release mode for every compilation job. If you want to have a closer point of comparison, you should try with a Red/System program (Red's low-level dialect):
```
Red/System [
   Title: "Simple hello world script"
]
print "Hello World!"
```
Save it as `hello.reds`, try to compile it and see the results. It's not as fast as gcc, but close enough. Given that Red/System compiler is entirely written in Rebol, which is a purely interpreted language (while gcc is written in C), the results are really decent actually.

**Comment by svnpenn** [comment]
@dockimbel whoa - that is totally better - under a second! couple of questions:

1. i assume i lose some functionality doing it that way?

2. doing it with `Red [`, can the Red Static Library be built by itself - this
way it wouldnt have to be done each time - the program itself could just be
compiled and linked with the Static Library

**Comment by endo64** [comment]
@svnpenn Please ask your questions on Gitter, in https://gitter.im/red/red/system or https://gitter.im/red/red or other related rooms.

**Comment by svnpenn** [comment]
@dockimbel whoa - that is totally better - under a second! couple of questions:

1. it seems using `Red/System [` you lose some functionality - for example this
script wont compile:

~~~red
Red [
   Title: "Simple hello world script"
]
print read http://example.com
~~~

2. doing it with `Red [`, can the Red Static Library be built by itself - this
way it wouldnt have to be compiled each time - the program itself could just be
compiled and linked with the Static Library

@endo64 you suggestion is not too helpful - between you and @PeterWAWood ive
seen these suggested:

- http://gitter.im/red/help
- http://gitter.im/red/red
- http://gitter.im/red/red/system

if you could narrow that to a single suggestion it would be more helpful.

**Comment by 9214** [comment]
@svnpenn 

1. I suggest you to read up on what is [Red](https://www.red-lang.org/p/about.html), what is [Red/System](https://static.red-lang.org/red-system-specs.html), and how one is [related to the other](https://github.com/red/red#red-programming-language) (the last paragraph).

2. What do you mean by 'Red Static Library'? Runtime? If your goal is to minimize compilation times, then re-read @PeterWAWood's answer. Another question is why you need to compile your program each time if Red is also an interpretive language.

And please, stop utilizing bug tracker for Q/A sessions. It was suggested to you multiple times to use Gitter chat for discussions. Either [`/help`](http://gitter.im/red/help) or [`/welcome`](https://gitter.im/red/red/welcome) room will do.

**Comment by svnpenn** [comment]
https://github.com/red/red/issues/3417

**Comment by 89z** [comment]
https://github.com/red/REP/issues/46

**Comment by 89z** [comment]
I agree with original poster, kind of strange not to offer static compile, as its offered by every major programming language today, such as Go, Nim, Rust. Anyone who prefers a static workflow has to wait 30 second every single time they compile anything.

**Comment by 9214** [comment]
@89z, aka @svnpenn, aka [Steve Penny](https://stackoverflow.com/users/1002260/steven-penny): nothing has changed since the team's [latest](https://github.com/red/red/issues/3417#issuecomment-396671976) encounter with you. If you want feature request to be implemented promptly and swiftly, just ask nicely and help with what you can afford.

**Comment by 89z** [comment]
@9214 your weird stalking aside, youre right. Nothing has changed. I saw the post on Hacker News, and wanted to put the word out for anyone who might be foolish enough to try this language. The fact that you have to compile the entire runtime every single time you want a static build, I dont see how anyone here can possibly think that is a good idea. All that would have to be done, is simply distribute a statically compiled runtime (`libred.a` or simliar), then it can be statically linked in a few seconds, instead of half a minute.

**Comment by greggirwin** [comment]
@89x, please be civil. @9214 isn't stalking you. You're dredging up old (closed) tickets in what seems like an effort to pick a fight. When you say "All that would have to be done...", you make it sound like it's something really easy to do. If it were, we would have done it. But it also hasn't been a priority.

**Comment by 89z** [comment]
> @89x, please be civil. @9214 isn't stalking you.

Nope. someone going around and posting link to various accounts of mine isnt appropriate behavior.

> You're dredging up old (closed) tickets in what seems like an effort to pick a fight.

The only interest I have in this ticket, is the fact the Red showed up on Hacker News. Notice carefully that I havent commented in several years.

> When you say "All that would have to be done...", you make it sound like it's something really easy to do. If it were, we would have done it. But it also hasn't been a priority.

I dont care about that. Its an objectively important factor in a programming language. If you dont agree with that, youre free to your opinion. However you should know that such decisions perhaps hold back the growth and adoption of the language.

**Comment by 9214** [comment]
@89z I am sorry if I made an impression of stalking you, which isn't the case. Your HN post had a recognizable polemic, and your *public* HN profile linked to a SO user with a familiar avatar and nickname. I merely connected the dots in my memory.

If your only interest in this ticket doesn't amount to contributing and discussing development in a constructive manner, perhaps it's time to lock the thread for good.

**Comment by 89z** [comment]
> @89z I am sorry if I made an impression of stalking you, which isn't the case. Your HN post had a recognizable polemic, and your _public_ HN profile linked to a SO user with a familiar avatar and nickname. I merely connected the dots in my memory.

You didnt need to do any of that. You stalked multiple profiles, and then even after that, you posted a pointless off topic comment [1], rather than actually address the point I made about compile time.

> If your only interest in this ticket doesn't amount to contributing and discussing development in a constructive manner, perhaps it's time to lock the thread for good.

Do whatever you want, I dont care. My point was to make very clear on Hacker News and here, what I consider a major shortcoming of the language. Ive done that.

1. https://news.ycombinator.com/item?id=27862463

**Comment by 89z** [comment]
Sanity check for myself:

~~~
PS C:\> cat main.go
package main
import "fmt"

func main() {
   fmt.Println("hello world")
}

PS C:\> Measure-Command { go build main.go }
Milliseconds      : 177
~~~

thats how it should be for static build, not 30 seconds.
